onnx_caffe2/backend.py

```diff
@@ -297,7 +299,65 @@ def _create_gemm(cls, n, opset_version):
         return ops
 
     @classmethod
-    def _create_pad(cls, n, opset_version):
+    def _create_lstm(cls, onnx_model, model_helper, n, opset_version):
+        assert model_helper is not None, "cannot convert LSTMs through this codepath"
+        assert onnx_model is not None, "cannot convert LSTMs through this codepath"
+        allowed_attrs = {'hidden_size'}
+        for x in n.attrs.keys():
+            assert x in allowed_attrs, "unsupported attribute " + x + " to LSTM"
+
+        input_blob, W, R, B, sequence_lens, initial_h, initial_c = n.inputs
+
+        hidden_size = n.attrs['hidden_size']
+
+        # ad-hoc, informally-specified, bug-ridden, slow
+        # implementation of shape inference
+        for x in onnx_model.graph.input:
+            if x.name == input_blob:
+                input_size = x.type.tensor_type.shape.dim[2].dim_value
+        for x in map(OnnxNode, onnx_model.graph.node):
+            if x.op_type == n.op_type and x.outputs[0] == input_blob:
+                input_size = x.attrs['hidden_size']
+
+        name = dummy_name()
+
+        # input and recurrence biases are squashed together in onnx but not in caffe2
```

Comment:
all of below should go to init_net probably (assuming that the weights are there). I guess you can check that inputs to weights are part of graph initializer.

Or is the plan to run 'staging' graph pass you have?

Reply:
unfortunately, this code runs after the optimizer, so it will not get pulled out into the init net. I could potentially put things into the init net explicitly, but that's more complicated (and, will these slices even show up as being expensive relative to the 'real' operations? I have no idea).

the pipeline is
ONNX ------(onnx-optimize)----->ONNX-------(this onnx-caffe2 code)------->Caffe2